org.jacoco.core/src/org/jacoco/core/internal/flow/Instruction.java

```diff
@@ -65,20 +69,35 @@ public void addBranch() {
 	 * @see #addBranch()
 	 * @param predecessor
 	 *            predecessor instruction
+	 * @param branch
+	 *            branch number in predecessor that should be marked as covered
+	 *            when this instruction marked as covered
 	 */
-	public void setPredecessor(final Instruction predecessor) {
+	public void setPredecessor(final Instruction predecessor,
```

Comment:
@marchof I also was puzzled by this multiple times.

Maybe terms "direct predecessor" and "in-between" are a bit vague: in case of
```
L0: IFEQ L2
L1: NOP
L2: NOP
```
`L1` will not be marked as predecessor of `L2` and there is probe "in-between" them
`L2` will be marked as predecessor of `L0`, however not so "direct" (i.e. not a previous instruction) - via jump and there is an "in-between" probe that is executed on this jump from `L0` to `L2` 😉 

But your proposal anyway better than nothing.

I updated javadoc and added check of invariant using a slightly modified version of your proposal.